### PR TITLE
[BugFix] fix num_lookahead_slots missing in async executor

### PR DESCRIPTION
**Pull request by leiwen83**
This fixes the following stacktrace:
```
handle: <Handle functools.partial(<function _raise_exception_on_finish at 0x7fed348f8d30>, error_callback=<bound method AsyncLLMEngine._error_callback of <vllm.engine.async_llm_engine.AsyncLLMEngine object at 0x7fed2a489ed0>>)>
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 38, in _raise_exception_on_finish
    task.result()
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 496, in run_engine_loop
    has_requests_in_progress = await asyncio.wait_for(
  File "/usr/lib/python3.10/asyncio/tasks.py", line 445, in wait_for
    return fut.result()
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 470, in engine_step
    request_outputs = await self.engine.step_async()
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 213, in step_async
    output = await self.model_executor.execute_model_async(
  File "/usr/local/lib/python3.10/dist-packages/vllm/executor/gpu_executor.py", line 152, in execute_model_async
    output = await make_async(self.driver_worker.execute_model)(
  File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
TypeError: SpecDecodeWorker.execute_model() missing 1 required positional argument: 'num_lookahead_slots'
```

### CODE REVIEW

**Comment by leiwen83** [comment]
cc @cadedaniel

**Review by rkooo567 — COMMENTED** [review]
Good catch. @cadedaniel is there any relevant test that could be updated to test it?

**Comment by cadedaniel** [comment]
Sorry I missed this. The fix looks good. To get it merged:
* Can we add a stacktrace with the failure exception to PR description?
* Can we add a test to catch this in the future?
  * Straightforward way is to add a test like https://github.com/vllm-project/vllm/blob/5434d905a6dd8828c34b1ff19c00a2c01ca687d3/tests/spec_decode/e2e/test_correctness.py#L69-L109, but which uses the AsyncLLMEngine instead of LLM entrypoint

**Comment by leiwen83** [comment]
Hi @cadedaniel ,

This issue is trigger while I am testing spec infer with openai api serving.
backtrace is as:
```
handle: <Handle functools.partial(<function _raise_exception_on_finish at 0x7fed348f8d30>, error_callback=<bound method AsyncLLMEngine._error_callback of <vllm.engine.async_llm_engine.AsyncLLMEngine object at 0x7fed2a489ed0>>)>
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 38, in _raise_exception_on_finish
    task.result()
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 496, in run_engine_loop
    has_requests_in_progress = await asyncio.wait_for(
  File "/usr/lib/python3.10/asyncio/tasks.py", line 445, in wait_for
    return fut.result()
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 470, in engine_step
    request_outputs = await self.engine.step_async()
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 213, in step_async
    output = await self.model_executor.execute_model_async(
  File "/usr/local/lib/python3.10/dist-packages/vllm/executor/gpu_executor.py", line 152, in execute_model_async
    output = await make_async(self.driver_worker.execute_model)(
  File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
TypeError: SpecDecodeWorker.execute_model() missing 1 required positional argument: 'num_lookahead_slots'
```

For testcase, I'm think whether we could add a new interface in vllm/entrypoints/llm.py, so that could make it as Async when request? Or we may need to implement similiar things in tests folder.

**Comment by cadedaniel** [comment]
> For testcase, I'm thinking whether we could add a new interface in vllm/entrypoints/llm.py, so that could make it as Async when request? Or we may need to implement similiar things in tests folder.

Exactly. In the ideal case we can run an async llm entrypoint, but that might be a lot of work -- it's ok to have a hackier version in a conftest somewher that allows us to get coverage of this codepath.

**Comment by leiwen83** [comment]
> > For testcase, I'm thinking whether we could add a new interface in vllm/entrypoints/llm.py, so that could make it as Async when request? Or we may need to implement similiar things in tests folder.
> 
> Exactly. In the ideal case we can run an async llm entrypoint, but that might be a lot of work -- it's ok to have a hackier version in a conftest somewher that allows us to get coverage of this codepath.

I try add use a AsyncEngine mode LLM implment in conftest, but found it would not free the cuda memory after delete. So it would broken following tests behind.
After a lot test, I find current we may need to resort to subprocess, and execute asyncllm inside it, while destory the subprocess itself after test finish. With this ugly build, now this error could be caught in test case and other normal syncllm case is also functional.

**Comment by zxdvd** [comment]
@leiwen83 CPU executor add async support recently, need to fix that too.

File: tests/spec_decode/e2e/conftest.py [hunk 1]
```diff
@@ -1,10 +1,132 @@
...
+Note: Crrurent LLM class in vllm don't support async mode, for test purpose,
+we implement async one in here. Maybe we could move to vllm/entrypoints/llm.py
+in future.
+
+Below AsyncLLM is directly borrow from vllm/entrypoints/llm.py with changes
+to make to work in async mode.
+"""
+
+
+class AsyncLLM:
```
Comment (rkooo567) [inline tests/spec_decode/e2e/conftest.py:30]: Consider supporting gpu cleanup using `cleanup` method in conftest.py? 

I think you can make it a conftest and delete the instance & call cleanup() in the destructor simliar to https://github.com/vllm-project/vllm/blob/468d761b32e3b3c5d64eeaa797e54ab809b7e50c/tests/conftest.py#L275

File: tests/spec_decode/e2e/test_correctness.py [hunk 2]
```diff
@@ -28,14 +28,116 @@
...
+    """Run generation with speculative decoding on a batch. Verify the engine
+    generates the correct number of tokens (via ignore_eos=True), and that the
+    detokenization matches HF transformers.
+    """
+    import time
+    from multiprocessing import Manager, Process
+
+    with Manager() as manager:
+        result = manager.dict()
+        p = Process(target=_test_spec_decode_e2e_with_detokenization_async,
```
Comment (rkooo567) [inline tests/spec_decode/e2e/test_correctness.py:131]: This is prone to leak. Consider;

```
try:
    p = Process(target=_test_spec_decode_e2e_with_detokenization_async,
                    args=(request, common_llm_kwargs,
                          per_test_common_llm_kwargs, test_llm_kwargs,
                          batch_size, seed, result))
    p.start()
    p.join()
finally:
    p.terminate()
```

or I recommend you to use ray (and use ray.shutdown() in the finally). 

```
import ray
try:
    ray.init()
    @ray.remote
    def run():
        # call your thing
    ray.get(run.remote())
finally:
    ray.shutdown()
```

File: tests/spec_decode/e2e/test_correctness.py [hunk 3]
```diff
@@ -28,14 +28,116 @@
...
+            for token_ids in batch_token_ids] == ([output_len] * batch_size)
+
+    # Expect detokenized string to match.
+    tok = AutoTokenizer.from_pretrained("JackFram/llama-68m")
+    for actual_tokens, actual_token_ids in zip(batch_tokens, batch_token_ids):
+        expected_tokens = tok.decode(actual_token_ids)
+        print(f"{actual_token_ids=}")
+        assert actual_tokens.strip() == expected_tokens.strip()
+
+    del llm
```
Comment (rkooo567) [inline tests/spec_decode/e2e/test_correctness.py:90]: Can you clean up as I suggested above ^?

**Review by rkooo567 — COMMENTED** [review]
Some comments regarding tests

**Comment by cadedaniel** [comment]
@rkooo567 's suggestions are good. Another way to do it is to look at spec decode e2e tests for inspiration:
https://github.com/vllm-project/vllm/blob/7923dcad12edae7dcfd6e0cf7ce2984b3bcecf0f/tests/spec_decode/e2e/conftest.py#L27-L52

**Comment by leiwen83** [comment]
> @rkooo567 's suggestions are good. Another way to do it is to look at spec decode e2e tests for inspiration:
> 
> https://github.com/vllm-project/vllm/blob/7923dcad12edae7dcfd6e0cf7ce2984b3bcecf0f/tests/spec_decode/e2e/conftest.py#L27-L52

simply delete llm and clean would not help reduce gpu memory usage here, which I think it is due to async mode creating a looping thread, and it would not be destory even llm itself is deleted.

I would try ray method.

File: tests/spec_decode/e2e/conftest.py [hunk 4]
```diff
@@ -1,10 +1,132 @@
...
 from vllm.model_executor.utils import set_random_seed
+from vllm.outputs import RequestOutput
+from vllm.sampling_params import SamplingParams
+from vllm.sequence import MultiModalData
+from vllm.usage.usage_lib import UsageContext
+from vllm.utils import Counter, random_uuid
+
+"""AsyncLLM
+
+Note: Crrurent LLM class in vllm don't support async mode, for test purpose,
```
[the remark this answers is not in the document]
Comment (leiwen83) [reply tests/spec_decode/e2e/conftest.py]: Got it, thanks for catching it.

**Comment by leiwen83** [comment]
> @leiwen83 CPU executor add async support recently, need to fix that too.

Thanks reminding, cpu execute also get fixed.

**Comment by leiwen83** [comment]
@rkooo567 now testcase is reworked with ray method.

@cadedaniel async mode in e2e case now could be compatible with original parameter passing way.

File: tests/spec_decode/e2e/conftest.py [hunk 5]
```diff
@@ -1,10 +1,143 @@
...
+        prompts: Optional[Union[str, List[str]]] = None,
+        sampling_params: Optional[Union[SamplingParams,
+                                        List[SamplingParams]]] = None,
+        prompt_token_ids: Optional[List[List[int]]] = None,
+        use_tqdm: bool = True,
+        lora_request: Optional[LoRARequest] = None,
+        multi_modal_data: Optional[MultiModalData] = None,
+    ) -> List[RequestOutput]:
+        try:
+            # make sure ray not init twice
```
Comment (rkooo567) [inline tests/spec_decode/e2e/conftest.py:85]: (you can also use ray.init(ignore_reinit_error =True))

File: tests/spec_decode/e2e/conftest.py [hunk 6]
```diff
@@ -1,10 +1,143 @@
...
+        tokenizer_revision: Optional[str] = None,
+        seed: int = 0,
+        gpu_memory_utilization: float = 0.9,
+        swap_space: int = 4,
+        enforce_eager: bool = False,
+        max_context_len_to_capture: int = 8192,
+        disable_custom_all_reduce: bool = False,
+        **kwargs,
+    ) -> None:
+        if "disable_log_stats" not in kwargs:
```
Comment (rkooo567) [inline tests/spec_decode/e2e/conftest.py:51]: Q: this implementation seems to be almost equivalent to `AsyncLLMEngine(engine_use_ray=True)`?

Comment (rkooo567) [reply tests/spec_decode/e2e/conftest.py:51]:
```
e = AsyncLLMEngine(engine_use_ray=True)
asyncio.run(e.generate(...))
```
?

**Review by rkooo567 — APPROVED** [review]
I think the current one is okay, but I have some impression we can just reuse `engine_use_ray` (lmk if this is wrong!)

File: tests/spec_decode/e2e/conftest.py (hunk 5 above)
Comment (leiwen83) [reply tests/spec_decode/e2e/conftest.py]: Apply~ thx

File: tests/spec_decode/e2e/conftest.py (hunk 6 above)
Comment (leiwen83) [reply tests/spec_decode/e2e/conftest.py:51]: with engine_use_ray=True and asyncio, it seems engine loop work still be alived after the run is called, which cause the next case oom ...

**Review by cadedaniel — APPROVED** [review]

**Review by cadedaniel — CHANGES_REQUESTED** [review]
Code looks good -- thanks!

what's up with the CI failure?

**Comment by leiwen83** [comment]
> Code looks good -- thanks!
> 
> what's up with the CI failure?

No idea... I rebase the code which retrigger the CI, seems still has similar issue, which looks like ray env not be cleaned up cause CI to hang waiting for something.

**Comment by leiwen83** [comment]
@cadedaniel I find the root cause of test case in CI failed. It is because test_spec_decode_xfail_ray in tests/spec_decode/e2e/test_compatibility.py has init the ray cluster which claim the gpu resource but never release it, causing the later ray call procedure fail to reclaim the gpu as there is only one GPU in the env.

**Review by cadedaniel — APPROVED** [review]

**Comment by andysalerno** [comment]
This may have introduced a failure on Dockerfile.cpu scenarios

```
python3 -m vllm.entrypoints.openai.api_server --model microsoft/Phi-3-mini-128k-instruct --trust-remote-code --max-model-len 8000
```

```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/uvicorn/protocols/http/httptools_impl.py", line 411, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/usr/local/lib/python3.10/dist-packages/uvicorn/middleware/proxy_headers.py", line 69, in __call__
    return await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/dist-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.10/dist-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.10/dist-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/usr/local/lib/python3.10/dist-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.10/dist-packages/starlette/middleware/cors.py", line 93, in __call__
    await self.simple_response(scope, receive, send, request_headers=headers)
  File "/usr/local/lib/python3.10/dist-packages/starlette/middleware/cors.py", line 148, in simple_response
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/dist-packages/starlette/middleware/exceptions.py", line 65, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py", line 756, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py", line 776, in app
    await route.handle(scope, receive, send)
  File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py", line 297, in handle
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py", line 77, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py", line 72, in app
    response = await func(request)
  File "/usr/local/lib/python3.10/dist-packages/fastapi/routing.py", line 278, in app
    raw_response = await run_endpoint_function(
  File "/usr/local/lib/python3.10/dist-packages/fastapi/routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
  File "/workspace/vllm/vllm/entrypoints/openai/api_server.py", line 90, in create_chat_completion
    generator = await openai_serving_chat.create_chat_completion(
  File "/workspace/vllm/vllm/entrypoints/openai/serving_chat.py", line 128, in create_chat_completion
    return await self.chat_completion_full_generator(
  File "/workspace/vllm/vllm/entrypoints/openai/serving_chat.py", line 290, in chat_completion_full_generator
    async for res in result_generator:
  File "/workspace/vllm/vllm/engine/async_llm_engine.py", line 663, in generate
    raise e
  File "/workspace/vllm/vllm/engine/async_llm_engine.py", line 657, in generate
    async for request_output in stream:
  File "/workspace/vllm/vllm/engine/async_llm_engine.py", line 77, in __anext__
    raise result
  File "/workspace/vllm/vllm/engine/async_llm_engine.py", line 38, in _raise_exception_on_finish
    task.result()
  File "/workspace/vllm/vllm/engine/async_llm_engine.py", line 498, in run_engine_loop
    has_requests_in_progress = await asyncio.wait_for(
  File "/usr/lib/python3.10/asyncio/tasks.py", line 445, in wait_for
    return fut.result()
  File "/workspace/vllm/vllm/engine/async_llm_engine.py", line 472, in engine_step
    request_outputs = await self.engine.step_async()
  File "/workspace/vllm/vllm/engine/async_llm_engine.py", line 213, in step_async
    output = await self.model_executor.execute_model_async(
  File "/workspace/vllm/vllm/executor/cpu_executor.py", line 114, in execute_model_async
    output = await make_async(self.driver_worker.execute_model)(
  File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
TypeError: CPUWorker.execute_model() got an unexpected keyword argument 'num_lookahead_slots'
```

**Comment by cadedaniel** [comment]
@andysalerno do you have a repro script?

**Comment by cadedaniel** [comment]
err, why is this not showing up in CI

**Pull request merged**